File: app/src/main/java/com/lumiyaviewer/lumiya/res/TextureCache.java
package com.lumiyaviewer.lumiya.res;
import java.util.*;

import android.util.Log;
import android.util.LruCache;
import com.lumiyaviewer.lumiya.memory.MemoryManager;
import com.lumiyaviewer.lumiya.memory.MemoryPressureListener;

/**
 * Texture cache that integrates with MemoryManager for proper memory management
 */
public class TextureCache implements MemoryPressureListener {
    private static final String TAG = "TextureCache";
    private static final int DEFAULT_CACHE_SIZE_MB = 32;
    
    private final MemoryManager memoryManager;
    private final LruCache<String, CachedTexture> cache;
    
    public TextureCache(MemoryManager memoryManager) {
        this.memoryManager = memoryManager;
        
        // Calculate cache size based on available memory
        Runtime runtime = Runtime.getRuntime();
        long maxMemory = runtime.maxMemory();
        int cacheSize = (int) (maxMemory / 8); // Use 1/8 of available memory
        
        this.cache = new LruCache<String, CachedTexture>(cacheSize) {
            @Override
            protected int sizeOf(String key, CachedTexture texture) {
                return texture.getEstimatedSize();
            }
            
            @Override
            protected void entryRemoved(boolean evicted, String key, 
                                       CachedTexture oldValue, CachedTexture newValue) {
                if (oldValue != null) {
                    memoryManager.trackDeallocation("texture_" + key, oldValue.getEstimatedSize());
                    oldValue.release();
                }
            }
        };
        
        memoryManager.addMemoryPressureListener(this);
    }
    
    public void put(String key, CachedTexture texture) {
        if (texture != null) {
            memoryManager.trackAllocation("texture_" + key, texture, texture.getEstimatedSize());
            cache.put(key, texture);
        }
    }
    
    public CachedTexture get(String key) {
        return cache.get(key);
    }
    
    public void remove(String key) {
        CachedTexture texture = cache.remove(key);
        if (texture != null) {
            memoryManager.trackDeallocation("texture_" + key, texture.getEstimatedSize());
            texture.release();
        }
    }
    
    public int getCacheSize() {
        return cache.size();
    }
    
    public void clear() {
        cache.evictAll();
    }
    
    @Override
    public void onMemoryPressure() {
        Log.i(TAG, "Memory pressure detected, trimming texture cache");
        // Trim to 50% of current size
        cache.trimToSize(cache.size() / 2);
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/res/TextureCache.kt
package com.lumiyaviewer.lumiya.res
import java.util.*

import android.util.Log
import android.util.LruCache
import com.lumiyaviewer.lumiya.memory.MemoryManager
import com.lumiyaviewer.lumiya.memory.MemoryPressureListener

/**
 * Texture cache that integrates with MemoryManager for proper memory management
 */
class TextureCache : MemoryPressureListener {
    private String TAG = "TextureCache";
    private Int DEFAULT_CACHE_SIZE_MB = 32
    
    private MemoryManager memoryManager
    private LruCache<String, CachedTexture> cache
    
    constructor(memoryManager: MemoryManager) {
        this.memoryManager = memoryManager
        
        // Calculate cache size based on available memory
        Runtime runtime = Runtime.getRuntime()
        Long maxMemory = runtime.maxMemory()
        Int cacheSize = (Int) (maxMemory / 8); // Use 1/8 of available memory
        
        this.cache = LruCache<String, CachedTexture>(cacheSize) {
            override protected fun sizeOf(key: String, texture: CachedTexture): Int {
                return texture.getEstimatedSize()
            }
            
            override protected fun entryRemoved(evicted: Boolean, key: String, oldValue: CachedTexture, newValue: CachedTexture): Unit {
                if (oldValue != null) {
                    memoryManager.trackDeallocation("texture_" + key, oldValue.getEstimatedSize());
                    oldValue.release()
                }
            }
        }
        
        memoryManager.addMemoryPressureListener(this)
    }
    
    fun put(key: String, texture: CachedTexture): Unit {
        if (texture != null) {
            memoryManager.trackAllocation("texture_" + key, texture, texture.getEstimatedSize());
            cache.put(key, texture)
        }
    }
    
    fun get(key: String): CachedTexture {
        return cache.get(key)
    }
    
    fun remove(key: String): Unit {
        CachedTexture texture = cache.remove(key)
        if (texture != null) {
            memoryManager.trackDeallocation("texture_" + key, texture.getEstimatedSize());
            texture.release()
        }
    }
    
    fun getCacheSize(): Int {
        return cache.size()
    }
    
    fun clear(): Unit {
        cache.evictAll()
    }
    
    override fun onMemoryPressure(): Unit {
        Log.i(TAG, "Memory pressure detected, trimming texture cache");
        // Trim to 50% of current size
        cache.trimToSize(cache.size() / 2)
    }
}
